pkg/oci/oci.go

```diff
@@ -272,26 +267,98 @@ func ListGadgetImages(ctx context.Context) ([]*GadgetImageDesc, error) {
 				log.Debugf("getting tag from image %q: %s", fullTag, err)
 				continue
 			}
-			imageColumn := &GadgetImageDesc{
+			image := &GadgetImageDesc{
 				Repository: repository,
 				Tag:        tag,
 			}
 
-			desc, err := ociStore.Resolve(ctx, fullTag)
+			desc, err := store.Resolve(ctx, fullTag)
 			if err != nil {
 				log.Debugf("Found tag %q but couldn't get a descriptor for it: %v", fullTag, err)
 				continue
 			}
-			imageColumn.Digest = desc.Digest.String()
-			imageColumns = append(imageColumns, imageColumn)
+			image.Digest = desc.Digest.String()
+			images = append(images, image)
 		}
 		return nil
 	})
+
+	return images, err
+}
+
+// ListGadgetImages lists all the gadget images.
+func ListGadgetImages(ctx context.Context) ([]*GadgetImageDesc, error) {
+	ociStore, err := getLocalOciStore()
+	if err != nil {
+		return nil, fmt.Errorf("getting oci store: %w", err)
+	}
+
+	images, err := listGadgetImages(ctx, ociStore)
 	if err != nil {
 		return nil, fmt.Errorf("listing all tags: %w", err)
 	}
 
-	return imageColumns, nil
+	return images, nil
+}
+
+func deleteTree(ctx context.Context, store *oci.Store, root ocispec.Descriptor) error {
+	predecessors, err := store.Predecessors(ctx, root)
+	if err != nil {
+		return fmt.Errorf("getting image predecessors: %w", err)
+	}
+
+	// We need to check if there are more than 1 predecessor as the current tree
+	// is counted in.
+	if len(predecessors) > 1 {
+		return nil
+	}
+
+	descriptors, err := content.Successors(ctx, store, root)
+	if err != nil {
+		return fmt.Errorf("getting image successors: %w", err)
+	}
+
+	for _, descriptor := range descriptors {
+		err := deleteTree(ctx, store, descriptor)
+		if err != nil {
+			return err
+		}
+	}
+
+	return store.Delete(ctx, root)
+}
+
+// DeleteGadgetImage removes the given image.
+func DeleteGadgetImage(ctx context.Context, image string) error {
+	ociStore, err := getLocalOciStore()
+	if err != nil {
+		return fmt.Errorf("getting oci store: %w", err)
+	}
+
+	targetImage, err := normalizeImageName(image)
+	if err != nil {
+		return fmt.Errorf("normalizing image: %w", err)
+	}
+
+	fullName := targetImage.String()
+	descriptor, err := ociStore.Resolve(ctx, fullName)
+	if err != nil {
+		return fmt.Errorf("resolving image: %w", err)
+	}
+
+	images, err := listGadgetImages(ctx, ociStore)
+	if err != nil {
+		return fmt.Errorf("listing images: %w", err)
+	}
+
+	digest := descriptor.Digest.String()
+	for _, img := range images {
+		imgFullName := fmt.Sprintf("%s:%s", img.Repository, img.Tag)
+		if img.Digest == digest && imgFullName != fullName {
+			return ociStore.Untag(ctx, fullName)
+		}
```

Comment:
Would you mind adding a comment indicating what's the reason for this check here? I think it's not so easy to understand. Something like:

`if there are multiple images pointing to the same digest, just untag the image instead of removing the whole tree`